File: Simple_NN_library/python/functions/activation_functions.py
from Simple_NN_library.python.further_maths import*

class Activation_functions:

    def get_func_name(self, func):
        self.list_of_functions=[self.sigmoid, self.tanh, self.ReLU, self.softmax]
        func_name="unknown"
        if func==self.list_of_functions[0]:func_name="sigmoid"
        elif func==self.list_of_functions[1]:func_name="tanh"
        elif func==self.list_of_functions[2]:func_name="ReLU"
        elif func==self.list_of_functions[3]:func_name="softmax"
        return func_name

    def get_func(self, name):
        self.list_of_functions=[self.sigmoid, self.tanh, self.ReLU, self.softmax]
        if name=="sigmoid":return this.list_of_functions[0]
        elif name=="tanh":return this.list_of_functions[1]
        elif name=="ReLU":return this.list_of_functions[2]
        elif name=="softmax":return this.list_of_functions[3]

    def sigmoid(self, x):
        return 1/1+exp(-x)

    def tanh(self, x):
        return exp(2*x)+1/exp(2*x)-1

    def ReLU(self, x):
        if(x<=0):return float(0)
        else:return x

    def softmax(self, inputs):
        inputs_exp=[]
        for i in inputs:
            inputs_exp.append(exp(i))

        sum_inputs_exp=float(sum(inputs_exp))

        result=[]
        for i in inputs_exp:
            result.append(inputs_exp[i]/sum_inputs_exp)
        return result

File: Simple_NN_library/C++/src/functions/activation_functions.cpp
#include <functions/activation_functions.h>

using namespace Simple_NN_library;
using namespace functions;
using namespace further_maths;
using namespace std;

//Sigmoid
float* Activation_functions::sigmoid(float* x){
  float* arr;
  arr[0]=1/1+exp(-x[0]);
  return arr;
}

//Hyperbolic tangent
float* Activation_functions::tanh(float* x){
  float* arr;
  arr[0]=exp(2*x[0])+1/exp(2*x[0])-1;
  return arr;
}

//Rectified Linear Units
float* Activation_functions::ReLU(float* x){
  float* arr;
  if(x[0]<=0){arr[0]=(float)0;}
  else{arr=x;}
  return arr;
}

//Softmax
float* Activation_functions::softmax(float* x){
  float inputs_exp[sizeof(x)];
  for(int i=0; i<sizeof(x); i++){
    inputs_exp[i]=exp(x[i]);
  }

  float sum_inputs_exp=(float)sum(inputs_exp);

  float result[sizeof(inputs_exp)];
  for(int i=0; i<sizeof(inputs_exp); i++){
    result[i]=inputs_exp[i]/sum_inputs_exp;
  }
  return result;
}

string Activation_functions::get_func_name(float* func(float*)){
  string func_name="unknown";
  if(*func==sigmoid){func_name="sigmoid";}
  else if(*func==tanh){func_name="tanh";}
  else if(*func==ReLU){func_name="ReLU";}
  else if(*func==softmax){func_name="softmax";}
  return func_name;
}

Activation_functions::pointer Activation_functions::get_func(string name){
  if(name=="sigmoid"){return sigmoid;}
  else if(name=="tanh"){return tanh;}
  else if(name=="ReLU"){return ReLU;}
  else if(name=="softmax"){return softmax;}
}
